# The Python Programming Language¶

The programming language you will be learning is Python. Python is an example of a high-level language; other high-level languages you might have heard of are C++, PHP, and Java.

As you might infer from the name high-level language, there are also low-level languages, sometimes referred to as machine languages or assembly languages. Machine language is the encoding of instructions in binary so that they can be directly executed by the computer. Assembly language uses a slightly easier format to refer to the low level instructions. Loosely speaking, computers can only execute programs written in low-level languages. To be exact, computers can actually only execute programs written in machine language. Thus, programs written in a high-level language (and even those in assembly language) have to be processed before they can run. This extra processing takes some time, which is a small disadvantage of high-level languages. However, the advantages to high-level languages are enormous.

First, it is much easier to program in a high-level language. Programs written in a high-level language take less time to write, they are shorter and easier to read, and they are more likely to be correct. Second, high-level languages are portable, meaning that they can run on different kinds of computers with few or no modifications. Low-level programs can run on only one kind of computer and have to be rewritten to run on another.

Due to these advantages, almost all programs are written in high-level languages. Low-level languages are used only for a few specialized applications.

Two kinds of programs process high-level languages into low-level languages: interpreters and compilers. An interpreter reads a high-level program and executes it, meaning that it does what the program says. It processes the program a little at a time, alternately reading lines and performing computations.

![](http://interactivepython.org/runestone/static/thinkcspy/_images/interpret.png)

A compiler reads the program and translates it completely before the program starts running. In this case, the high-level program is called the source code, and the translated program is called the object code or the executable. Once a program is compiled, you can execute it repeatedly without further translation. 	

![](http://interactivepython.org/runestone/static/thinkcspy/_images/compile.png)

Many modern languages use both processes. They are first compiled into a lower level language, called byte code, and then interpreted by a program called a virtual machine. Python uses both processes, but because of the way programmers interact with it, it is usually considered an interpreted language.

There are two ways to use the Python interpreter: shell mode and program mode. In shell mode, you type Python expressions into the Python shell, and the interpreter immediately shows the result. The example below shows the Python shell at work.

<pre>
'''python
$ python3
Python 3.2 (r32:88445, Mar 25 2011, 19:28:28)
[GCC 4.5.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> 2 + 3
5
>>>
'''

</pre>


The >>> is called the Python prompt. The interpreter uses the prompt to indicate that it is ready for instructions. We typed 2 + 3. The interpreter evaluated our expression and replied 5. On the next line it gave a new prompt indicating that it is ready for more input.

Working directly in the interpreter is convenient for testing short bits of code because you get immediate feedback. Think of it as scratch paper used to help you work out problems.


Alternatively, you can write an entire program by placing lines of Python instructions in a file and then use the interpreter to execute the contents of the file as a whole. Such a file is often referred to as source code. For example, we used a text editor to create a source code file named firstprogram.py with the following contents:

In [1]:
print("My first program adds two numbers, 2 and 3:")
print(2 + 3)

My first program adds two numbers, 2 and 3:
5


By convention, files that contain Python programs have names that end with .py . Following this convention will help your operating system and other programs identify a file as containing python code.

<pre>
'''python
$ python firstprogram.py

My first program adds two numbers, 2 and 3:
5
'''
</pre>

These examples show Python being run from a Unix command line. In other development environments, the details of executing programs may differ. Also, most programs are more interesting than this one.

Want to learn more about Python?
If you would like to learn more about installing and using Python, here are some video links. Installing Python for Windows shows you how to install the Python environment under Windows Vista, Installing Python for Mac shows you how to install under Mac OS/X, and Installing Python for Linux shows you how to install from the Linux command line. Using Python shows you some details about the Python shell and source code.

## Check your understanding

1: Source code is another name for:
(A) the instructions in a program, stored in a file.
(B) the language that you are programming in (e.g., Python).
(C) the environment/tool in which you are programming.
(D) the number (or "code") that you must input at the top of each program to tell the computer how to execute your program.


2: What is the difference between a high-level programming language and a low-level programming language?
(A) It is high-level if you are standing and low-level if you are sitting.
(B) It is high-level if you are programming for a computer and low-level if you are programming for a phone or mobile device.
(C) It is high-level if the program must be processed before it can run, and low-level if the computer can execute it without additional processing.
(D) It is high-level if it easy to program in and is very short; it is low-level if it is really hard to program in and the programs are really long.


3: Pick the best replacements for 1 and 2 in the following sentence: When comparing compilers and interpreters, a compiler is like 1 while an interpreter is like 2.
(A) 1 = a process, 2 = a function
(B) 1 = translating an entire book, 2 = translating a line at a time
(C) 1 = software, 2 = hardware
(D) 1 = object code, 2 = byte code

# Special Ways To Execute This Book

This book provides two special ways to execute Python programs. Both techniques are designed to assist you as you learn the Python programming language. They will help you increase your understanding of how Python programs work.

First, you can write, modify, and execute programs using a unique activecode interpreter that allows you to execute Python code right in the text itself (right from the web browser). Although this is certainly not the way real programs are written, it provides an excellent environment for learning a programming language like Python since you can experiment with the language as you are reading.

Take a look at the activecode interpreter in action. If we use the Python code from the previous example and make it active, you will see that it can be executed directly by pressing the run button. Try pressing the run button below.

In [2]:
print("My first program adds two numbers, 2 and 3:")

print(2 + 3)

My first program adds two numbers, 2 and 3:
5


Now try modifying the activecode program shown above. First, modify the string in the first print statement by changing the word adds to the word multiplies. Now press run. You can see that the result of the program has changed. However, it still prints “5” as the answer. Modify the second print statement by changing the addition symbol, the “+”, to the multiplication symbol, “*”. Press run to see the new results.


In [3]:
print("My first program adds two numbers, 2 and 3:")
print(2 + 3)

My first program adds two numbers, 2 and 3:
5


# More About Programs

A program is a sequence of instructions that specifies how to perform a computation. The computation might be something as complex as rendering an html page in a web browser or encoding a video and streaming it across the network. It can also be a symbolic computation, such as searching for and replacing text in a document or (strangely enough) compiling a program.

The details look different in different languages, but a few basic instructions appear in just about every language.

**input**<br>
Get data from the keyboard, a file, or some other device.

**output**<br>
Display data on the screen or send data to a file or other device.

**math** and **logic**<br>
Perform basic mathematical operations like addition and multiplication and logical operations like and, or, and not.
conditional execution
Check for certain conditions and execute the appropriate sequence of statements.

**repetition**<br>
Perform some action repeatedly, usually with some variation.
Believe it or not, that’s pretty much all there is to it. Every program you’ve ever used, no matter how complicated, is made up of instructions that look more or less like these. Thus, we can describe programming as the process of breaking a large, complex task into smaller and smaller subtasks until the subtasks are simple enough to be performed with sequences of these basic instructions.


## Check your understanding

A program is:<br>
(A) a sequence of instructions that specifies how to perform a computation. <br>
(B) something you follow along at a play or concert.<br>
(C) a computation, even a symbolic computation.<br>
(D) the same thing as an algorithm.<br>


## What is Debugging?

Programming is a complex process. Since it is done by human beings, errors may often occur. Programming errors are called bugs and the process of tracking them down and correcting them is called debugging. Some claim that in 1945, a dead moth caused a problem on relay number 70, panel F, of one of the first computers at Harvard, and the term bug has remained in use since. For more about this historic event, see first bug.

Three kinds of errors can occur in a program: syntax errors, runtime errors, and semantic errors. It is useful to distinguish between them in order to track them down more quickly.

## Check your understanding

Debugging is: <br>
(A) tracking down programming errors and correcting them.<br>
(B) removing all the bugs from your house.<br>
(C) finding all the bugs in the program.<br>
(D) fixing the bugs in the program.<br>


## Syntax errors
Python can only execute a program if the program is syntactically correct; otherwise, the process fails and returns an error message. Syntax refers to the structure of a program and the rules about that structure. For example, in English, a sentence must begin with a capital letter and end with a period. this sentence contains a syntax error. So does this one

For most readers, a few syntax errors are not a significant problem, which is why we can read the poetry of e. e. cummings without problems. Python is not so forgiving. If there is a single syntax error anywhere in your program, Python will display an error message and quit. You will not be able to complete the execution of your program. During the first few weeks of your programming career, you will probably spend a lot of time tracking down syntax errors. However, as you gain experience, you will make fewer errors and you will also be able to find your errors faster.

## Check your understanding

Which of the following is a syntax error?<br>
(A) Attempting to divide by 0.<br>
(B) Forgetting a colon at the end of a statement where one is required.<br>
(C) Forgetting to divide by 100 when printing a percentage amount.<br>


Who or what typically finds syntax errors?<br>
(A) The programmer.<br>
(B) The compiler / interpreter.<br>
(C) The computer.<br>
(D) The teacher / instructor. <br>

## Runtime Errors
The second type of error is a runtime error, so called because the error does not appear until you run the program. These errors are also called exceptions because they usually indicate that something exceptional (and bad) has happened.

Runtime errors are rare in the simple programs you will see in the first few chapters, so it might be a while before you encounter one.


## Check your understanding

Which of the following is a run-time error? <br>
(A) Attempting to divide by 0.<br>
(B) Forgetting a colon at the end of a statement where one is required.<br>
(C) Forgetting to divide by 100 when printing a percentage amount.<br>


## Semantic Errors
The third type of error is the semantic error. If there is a semantic error in your program, it will run successfully in the sense that the computer will not generate any error messages. However, your program will not do the right thing. It will do something else. Specifically, it will do what you told it to do.

The problem is that the program you wrote is not the program you wanted to write. The meaning of the program (its semantics) is wrong. Identifying semantic errors can be tricky because it requires you to work backward by looking at the output of the program and trying to figure out what it is doing.


## Check your understanding

Which of the following is a semantic error?<br>
(A) Attempting to divide by 0.<br>
(B) Forgetting a semi-colon at the end of a statement where one is required.<br>
(C) Forgetting to divide by 100 when printing a percentage amount.


## Experimental Debugging
One of the most important skills you will acquire is debugging. Although it can be frustrating, debugging is one of the most intellectually rich, challenging, and interesting parts of programming.

In some ways, debugging is like detective work. You are confronted with clues, and you have to infer the processes and events that led to the results you see.

Debugging is also like an experimental science. Once you have an idea what is going wrong, you modify your program and try again. If your hypothesis was correct, then you can predict the result of the modification, and you take a step closer to a working program. If your hypothesis was wrong, you have to come up with a new one. As Sherlock Holmes pointed out, When you have eliminated the impossible, whatever remains, however improbable, must be the truth. (A. Conan Doyle, The Sign of Four)

For some people, programming and debugging are the same thing. That is, programming is the process of gradually debugging a program until it does what you want. The idea is that you should start with a program that does something and make small modifications, debugging them as you go, so that you always have a working program.

For example, Linux is an operating system kernel that contains millions of lines of code, but it started out as a simple program Linus Torvalds used to explore the Intel 80386 chip. According to Larry Greenfield, one of Linus’s earlier projects was a program that would switch between displaying AAAA and BBBB. This later evolved to Linux (The Linux Users’ Guide Beta Version 1).

Later chapters will make more suggestions about debugging and other programming practices.


## Check your understanding

The difference between programming and debugging is:<br>
(A) programming is the process of writing and gradually debugging a program until it does what you want.<br>
(B) programming is creative and debugging is routine.<br>
(C) programming is fun and debugging is work.<br>
(D) there is no difference between them.<br>

## A Typical First Program

Traditionally, the first program written in a new language is called Hello, World! because all it does is display the words, Hello, World! In Python, the source code looks like this.

In [5]:
print("Hello, World!")

Hello, World!


This is an example of using the print function, which doesn’t actually print anything on paper. It displays a value on the screen. In this case, the result is the phrase:

<pre>
Hello, World!
</pre>


Now you try printing something! 

In [6]:
print("Hello, World!")

Hello, World!


The quotation marks in the program mark the beginning and end of the value. They don’t appear in the result.

Some people judge the quality of a programming language by the simplicity of the Hello, World! program. By this standard, Python does about as well as possible.


## Check your understanding

The print function:<br>
(A) sends information to the printer to be printed on paper.
(B) displays a value on the screen.<br>
(C) tells the computer to put the information in print, rather than cursive, format.<br>
(D) tells the computer to speak the information.<br>


## Comments
As programs get bigger and more complicated, they get more difficult to read. Formal languages are dense, and it is often difficult to look at a piece of code and figure out what it is doing, or why. For this reason, it is a good idea to add notes to your programs to explain in natural language what the program is doing. These notes are called comments.

A comment in a computer program is text that is intended only for the human reader - it is completely ignored by the interpreter. In Python, the # token starts a comment. The rest of the line is ignored. Here is a new version of Hello, World!.

In [7]:
#---------------------------------------------------
# This demo program shows off how elegant Python is!
# Written by Scarlet, June 11.
# Anyone may freely copy or modify this program.
#---------------------------------------------------

print("Hello, World!")     # Isn't this easy!


Hello, World!


Notice that when you run this program, it still only prints the phrase Hello, World! None of the comments appear. You’ll also notice that we’ve left a blank line in the program. Blank lines are also ignored by the interpreter, but comments and blank lines can make your programs much easier for humans to parse. Use them liberally!



## Check your understanding

What are comments for?<br>
(A) To tell the computer what you mean in your program.<br>
(B) For the people who are reading your code to know, in natural language, what the program is doing.<br>
(C) Nothing, they are extraneous information that is not needed.<br>
(D) Nothing in a short program. They are only needed for really large programs.<br>

## Values and Data Types
A value is one of the fundamental things — like a word or a number — that a program manipulates. The values we have seen so far are 5 (the result when we added 2 + 3), and "Hello, World!". We often refer to these values as objects and we will use the words value and object interchangeably.

### Note
Actually, the 2 and the 3 that are part of the addition above are values(objects) as well.
These objects are classified into different classes, or data types: 4 is an integer, and "Hello, World!" is a string, so-called because it contains a string or sequence of letters. You (and the interpreter) can identify strings because they are enclosed in quotation marks.

If you are not sure what class a value falls into, Python has a function called type which can tell you.

In [8]:
print(type("Hello, World!"))
print(type(17))
print("Hello, World")


<class 'str'>
<class 'int'>
Hello, World


Not surprisingly, strings belong to the class str and integers belong to the class int.

## Note
When we show the value of a string using the print function, such as in the third example above, the quotes are no longer present. The value of the string is the sequence of characters inside the quotes. The quotes are only necessary to help Python know what the value is.
In the Python shell, it is not necessary to use the print function to see the values shown above. The shell evaluates the Python function and automatically prints the result. For example, consider the shell session shown below. When we ask the shell to evaluate type("Hello, World!"), it responds with the appropriate answer and then goes on to display the prompt for the next use.

<pre>
Python 3.1.2 (r312:79360M, Mar 24 2010, 01:33:18)
[GCC 4.0.1 (Apple Inc. build 5493)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> type("Hello, World!")
<class 'str'>
>>> type(17)
<class 'int'>
>>> "Hello, World"
'Hello, World'
>>>
</pre>

> Note that in the last example, we simply ask the shell to evaluate the string “Hello, World”. The result is as you might expect, the string itself.

Continuing with our discussion of data types, numbers with a decimal point belong to a class called float, because these numbers are represented in a format called floating-point. At this stage, you can treat the words class and type interchangeably. We’ll come back to a deeper understanding of what a class is in later chapters.

In [9]:
print(type(3.2))


<class 'float'>


What about values like "17" and "3.2"? They look like numbers, but they are in quotation marks like strings.

In [10]:
print(type("17"))
print(type("3.2"))


<class 'str'>
<class 'str'>


They’re strings!

Strings in Python can be enclosed in either single quotes (') or double quotes ("), or three of each (''' or """)

In [11]:
print(type('This is a string.') )
print(type("And so is this.") )
print(type("""and this.""") )
print(type('''and even this...''') )


<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>


Double quoted strings can contain single quotes inside them, as in "Bruce's beard", and single quoted strings can have double quotes inside them, as in 'The knights who say "Ni!"'. Strings enclosed with three occurrences of either quote symbol are called triple quoted strings. They can contain either single or double quotes:




In [12]:
print('''"Oh no", she exclaimed, "Ben's bike is broken!"''')


"Oh no", she exclaimed, "Ben's bike is broken!"


Triple quoted strings can even span multiple lines:

In [13]:
message = """This message will
span several
lines."""
print(message)

print("""This message will span
several lines
of the text.""")


This message will
span several
lines.
This message will span
several lines
of the text.


Python doesn’t care whether you use single or double quotes or the three-of-a-kind quotes to surround your strings. Once it has parsed the text of your program or command, the way it stores the value is identical in all cases, and the surrounding quotes are not part of the value.



In [14]:
print('This is a string.')
print("""And so is this.""")


This is a string.
And so is this.


So the Python language designers usually chose to surround their strings by single quotes. What do you think would happen if the string already contained single quotes?

When you type a large integer, you might be tempted to use commas between groups of three digits, as in <pre>42,000</pre> This is not a legal integer in Python, but it does mean something else, which is legal:

In [16]:
print(42000)
print(42,000)


42000
42 0


Well, that’s not what we expected at all! Because of the comma, Python chose to treat this as a pair of values. In fact, the print function can print any number of values as long as you separate them by commas. Notice that the values are separated by spaces when they are displayed.

In [17]:
print(42, 17, 56, 34, 11, 4.35, 32)
print(3.4, "hello", 45)


42 17 56 34 11 4.35 32
3.4 hello 45


Remember not to put commas or spaces in your integers, no matter how big they are. Also revisit what we said in the previous chapter: formal languages are strict, the notation is concise, and even the smallest change might mean something quite different from what you intended.



## Check your understanding

How can you determine the type of a variable?<br>
(A) Print out the value and determine the data type based on the value printed.<br>
(B) Use the type function.<br>
(C) Use it in a known equation and print the result.<br>
(D) Look at the declaration of the variable.<br>
<br>
What is the data type of ‘this is what kind of data’?<br>
(A) Character<br>
(B) Integer<br>
(C) Float<br>
(D) String<br>

## Type conversion functions
Sometimes it is necessary to convert values from one type to another. Python provides a few simple functions that will allow us to do that. The functions <pre>int</pre>, <pre>float</pre> and <pre>str</pre> will (attempt to) convert their arguments into types int, float and str respectively. We call these **type conversion** functions.

The <pre>int()</pre> function can take a floating point number or a string, and turn it into an int. For floating point numbers, it discards the decimal portion of the number - a process we call truncation towards zero on the number line. Let us see this in action:




In [20]:
print(3.14, int(3.14))
print(3.9999, int(3.9999))       # This doesn't round to the closest int!
print(3.0, int(3.0))
print(-3.999, int(-3.999))        # Note that the result is closer to zero

print("2345", int("2345"))        # parse a string to produce an int
print(17, int(17))                # int even works on integers
print(int("23bottles"))


3.14 3
3.9999 3
3.0 3
-3.999 -3
2345 2345
17 17


ValueError: invalid literal for int() with base 10: '23bottles'

### Error
#### ValueError: 
invalid literal for int() with base 10: '23bottles' on line 8

#### Description
A ValueError most often occurs when you pass a parameter to a function and the function is expecting one type and you pass another.

#### To Fix
The error message gives you a pretty good hint about the name of the function as well as the value that is incorrect. Look at the error message closely and then trace back to the variable containing the problematic value.

The last case shows that a string has to be a syntactically legal number, otherwise you’ll get one of those pesky runtime errors. Modify the example by deleting the <pre>bottles</pre> and rerun the program. You should see the integer <pre>23</pre>

The type converter <pre>float</pre> can turn an integer, a float, or a syntactically legal string into a float.

In [21]:
print(float("123.45"))
print(type(float("123.45")))


123.45
<class 'float'>


The type converter <pre>str</pre> turns its argument into a string. Remember that when we print a string, the quotes are removed. However, if we print the type, we can see that it is definitely str.




In [22]:
print(str(17))
print(str(123.45))
print(type(str(123.45)))

17
123.45
<class 'str'>


## Check your understanding

What value is printed when the following statement executes?

<pre>print( int(53.785) )</pre>
(A) Nothing is printed. It generates a runtime error.<br>
(B) 53<br>
(C) 54<br>
(D) 53.785<br>

## Variables

One of the most powerful features of a programming language is the ability to manipulate variables. A variable is a name that refers to a value.

**Assignment statements** create new variables and also give them values to refer to.

In [23]:
message = "What's up, Doc?"
n = 17
pi = 3.14159

***marked codecolor***

This example makes three assignments. The first assigns the string value "What's up, Doc?" to a new variable named message. The second gives the integer 17 to n, and the third assigns the floating-point number 3.14159 to a variable called pi.

The assignment token, =, should not be confused with equality (we will see later that equality uses the == token). The assignment statement links a name, on the left hand side of the operator, with a value, on the right hand side. This is why you will get an error if you enter:

17 = n

> **Tip**<br>
When reading or writing code, say to yourself “n is assigned 17” or “n gets the value 17” or “n is a reference to the object 17” or “n refers to the object 17”. Don’t say “n equals 17”.

A common way to represent variables on paper is to write the name with an arrow pointing to the variable’s value. This kind of figure, known as a reference diagram, is often called a state snapshot because it shows what state each of the variables is in at a particular instant in time. (Think of it as the variable’s state of mind). This diagram shows the result of executing the assignment statements shown above.

![](http://interactivepython.org/runestone/static/thinkcspy/_images/refdiagram1.png)

If you ask Python to evaluate a variable, it will produce the value that is currently linked to the variable. In other words, evaluating a variable will give you the value that is referred to by the variable.

If you ask Python to evaluate a variable, it will produce the value that is currently linked to the variable. In other words, evaluating a variable will give you the value that is referred to by the variable.

In [25]:
message = "What's up, Doc?"
n = 17
pi = 3.14159

print(message)
print(n)
print(pi)


What's up, Doc?
17
3.14159


Variables also have types; again, we can ask the interpreter what they are.

In [26]:
message = "What's up, Doc?"
n = 17
pi = 3.14159

print(type(message))
print(type(n))
print(type(pi))


<class 'str'>
<class 'int'>
<class 'float'>


The type of a variable is the type of the object it currently refers to.

We use variables in a program to “remember” things, like the current score at the football game. But variables are variable. This means they can change over time, just like the scoreboard at a football game. You can assign a value to a variable, and later assign a different value to the same variable.

> **Note**
This is different from math. In math, if you give x the value 3, it cannot change to refer to a different value half-way through your calculations!

To see this, read and then run the following program. You’ll notice we change the value of day three times, and on the third assignment we even give it a value that is of a different type.

In [27]:
day = "Thursday"
print(day)
day = "Friday"
print(day)
day = 21
print(day)

Thursday
Friday
21


A great deal of programming is about having the computer remember things. For example, we might want to keep track of the number of missed calls on your phone. Each time another call is missed, we will arrange to update or change the variable so that it will always reflect the correct value.

## Check your understanding

What is printed when the following statements execute?<br>
day = "Thursday"<br>
day = 32.5<br>
day = 19<br>
print(day)<br><br>
(A) Nothing is printed. A runtime error occurs.<br>
(B) Thursday<br>
(C) 32.5<br>
(D) 19<br>

## Variable Names and Keywords
Variable names can be arbitrarily long. They can contain both letters and digits, but they have to begin with a letter or an underscore. Although it is legal to use uppercase letters, by convention we don’t. If you do, remember that case matters. Bruce and bruce are different variables.

> Caution
Variable names can never contain spaces.

The underscore character ( _) can also appear in a name. It is often used in names with multiple words, such as my_name or price_of_tea_in_china. There are some situations in which names beginning with an underscore have special meaning, so a safe rule for beginners is to start all names with a letter.

If you give a variable an illegal name, you get a syntax error. In the example below, each of the variable names is illegal.



In [28]:
76trombones = "big parade"
more$ = 1000000
class = "Computer Science 101"

SyntaxError: invalid syntax (<ipython-input-28-f6adc31ec82a>, line 1)

76trombones is illegal because it does not begin with a letter. more$ is illegal because it contains an illegal character, the dollar sign. But what’s wrong with class?

It turns out that class is one of the Python keywords. Keywords define the language’s syntax rules and structure, and they cannot be used as variable names. Python has thirty-something keywords (and every now and again improvements to Python introduce or eliminate one or two):

and	as	assert	break	class	continue
def	del	elif	else	except	exec
finally	for	from	global	if	import
in	is	lambda	nonlocal	not	or
pass	raise	return	try	while	with
yield	True	False	None	 	


You might want to keep this list handy. If the interpreter complains about one of your variable names and you don’t know why, see if it is on this list.

Programmers generally choose names for their variables that are meaningful to the human readers of the program — they help the programmer document, or remember, what the variable is used for.

Caution
Beginners sometimes confuse “meaningful to the human readers” with “meaningful to the computer”. So they’ll wrongly think that because they’ve called some variable average or pi, it will somehow automagically calculate an average, or automagically associate the variable pi with the value 3.14159. No! The computer doesn’t attach semantic meaning to your variable names.

So you’ll find some instructors who deliberately don’t choose meaningful names when they teach beginners — not because they don’t think it is a good habit, but because they’re trying to reinforce the message that you, the programmer, have to write some program code to calculate the average, or you must write an assignment statement to give a variable the value you want it to have.



## Check your understanding

True or False:<br> 
the following is a legal variable name in Python: <br>
A_good_grade_is_A+<br>
(A) True<br>
(B) False<br>

## Statements and Expressions

A statement is an instruction that the Python interpreter can execute. We have only seen the assignment statement so far. Some other kinds of statements that we’ll see shortly are while statements, for statements, if statements, and import statements. (There are other kinds too!)

An expression is a combination of values, variables, operators, and calls to functions. Expressions need to be evaluated. If you ask Python to print an expression, the interpreter evaluates the expression and displays the result.

In [29]:
print(1 + 1)
print(len("hello"))


2
5


In this example len is a built-in Python function that returns the number of characters in a string. We’ve previously seen the print and the type functions, so this is our third example of a function!

The evaluation of an expression produces a value, which is why expressions can appear on the right hand side of assignment statements. A value all by itself is a simple expression, and so is a variable. Evaluating a variable gives the value that the variable refers to.

In [30]:
y = 3.14
x = len("hello")
print(x)
print(y)


5
3.14


If we take a look at this same example in the Python shell, we will see one of the distinct differences between statements and expressions.

<pre>
'''
>>> y = 3.14
>>> x = len("hello")
>>> print(x)
5
>>> print(y)
3.14
>>> y
3.14
>>>
'''
</pre>

Note that when we enter the assignment statement, y = 3.14, only the prompt is returned. There is no value. This is due to the fact that statements, such as the assignment statement, do not return a value. They are simply executed.

On the other hand, the result of executing the assignment statement is the creation of a reference from a variable, y, to a value, 3.14. When we execute the print function working on y, we see the value that y is referring to. In fact, evaluating y by itself results in the same response.

## Operators and Operands
Operators are special tokens that represent computations like addition, multiplication and division. The values the operator works on are called operands.

The following are all legal Python expressions whose meaning is more or less clear:

<pre>
'''
20 + 32
hour - 1
hour * 60 + minute
minute / 60
5 ** 2
(5 + 9) * (15 - 7)
'''
</pre>
The tokens +, -, and *, and the use of parenthesis for grouping, mean in Python what they mean in mathematics. The asterisk (*) is the token for multiplication, and ** is the token for exponentiation. Addition, subtraction, multiplication, and exponentiation all do what you expect.

In [31]:
print(2 + 3)
print(2 - 3)
print(2 * 3)
print(2 ** 3)
print(3 ** 2)


5
-1
6
8
9


When a variable name appears in the place of an operand, it is replaced with the value that it refers to before the operation is performed. For example, what if we wanted to convert 645 minutes into hours. In Python 3, division is denoted by the operator token / which always evaluates to a floating point result.

In [32]:
minutes = 645
hours = minutes / 60
print(hours)


10.75


What if, on the other hand, we had wanted to know how many whole hours there are and how many minutes remain. To help answer this question, Python gives us a second flavor of the division operator. This version, called integer division, uses the token //. It always truncates its result down to the next smallest integer (to the left on the number line).

In [33]:
print(7 / 4)
print(7 // 4)
minutes = 645
hours = minutes // 60
print(hours)


1.75
1
10


Pay particular attention to the first two examples above. Notice that the result of floating point division is 1.75 but the result of the integer division is simply 1. Take care that you choose the correct flavor of the division operator. If you’re working with expressions where you need floating point values, use the division operator /. If you want an integer result, use //.

The modulus operator, sometimes also called the remainder operator or integer remainder operator works on integers (and integer expressions) and yields the remainder when the first operand is divided by the second. In Python, the modulus operator is a percent sign (%). The syntax is the same as for other operators.

In [34]:
quotient = 7 // 3     # This is the integer division operator
print(quotient)
remainder = 7 % 3
print(remainder)


2
1


In the above example, 7 divided by 3 is 2 when we use integer division and there is a remainder of 1.

The modulus operator turns out to be surprisingly useful. For example, you can check whether one number is divisible by another—if x % y is zero, then x is divisible by y. Also, you can extract the right-most digit or digits from a number. For example, x % 10 yields the right-most digit of x (in base 10). Similarly x % 100 yields the last two digits.

Finally, returning to our time example, the remainder operator is extremely useful for doing conversions, say from seconds, to hours, minutes and seconds. If we start with a number of seconds, say 7684, the following program uses integer division and remainder to convert to an easier form. Step through it to be sure you understand how the division and remainder operators are being used to compute the correct values.

## Check your understanding

What value is printed when the following statement executes?

print(18 / 4)<br>

(A) 4.5<br>
(B) 5<br>
(C) 4<br>
(D) 2<br>




What value is printed when the following statement executes?
print(18 // 4)<br>
(A) 4.25<br>
(B) 5<br>
(C) 4<br>
(D) 2<br>



What value is printed when the following statement executes?
print(18 % 4)<br>
(A) 4.25<br>
(B) 5<br>
(C) 4<br>
(D) 2<br>

## Input

The program in the previous section works fine but is very limited in that it only works with one value for total_secs. What if we wanted to rewrite the program so that it was more general. One thing we could do is allow the user to enter any value they wish for the number of seconds. The program could then print the proper result for that starting value.

In order to do this, we need a way to get input from the user. Luckily, in Python there is a built-in function to accomplish this task. As you might expect, it is called input.

<pre>n = input("Please enter your name: ")</pre>

The input function allows the user to provide a prompt string. When the function is evaluated, the prompt is shown. The user of the program can enter the name and press return. When this happens the text that has been entered is returned from the input function, and in this case assigned to the variable n. Make sure you run this example a number of times and try some different names in the input box that appears.

RunLoad HistoryShow CodeLens


In [35]:
n = input("Please enter your name: ")
print("Hello", n)


Please enter your name: sdfsd
Hello sdfsd


It is very important to note that the input function returns a string value. Even if you asked the user to enter their age, you would get back a string like "17". It would be your job, as the programmer, to convert that string into an int or a float, using the int or float converter functions we saw earlier.

To modify our previous program, we will add an input statement to allow the user to enter the number of seconds. Then we will convert that string to an integer. From there the process is the same as before. To complete the example, we will print some appropriate output.




In [36]:
str_seconds = input("Please enter the number of seconds you wish to convert")
total_secs = int(str_seconds)

hours = total_secs // 3600
secs_still_remaining = total_secs % 3600
minutes =  secs_still_remaining // 60
secs_finally_remaining = secs_still_remaining  % 60

print("Hrs=", hours, "mins=", minutes, "secs=", secs_finally_remaining)


Please enter the number of seconds you wish to convert234
Hrs= 0 mins= 3 secs= 54


The variable str_seconds will refer to the string that is entered by the user. As we said above, even though this string may be 7684, it is still a string and not a number. To convert it to an integer, we use the int function. The result is referred to by total_secs. Now, each time you run the program, you can enter a new value for the number of seconds to be converted.

## Check your understanding

What is printed when the following statements execute?<br>
<pre>
''' 
n = input("Please enter your age: ")?<br>
     ***user types in 18***
print ( type(n) )
'''
</pre><br>
(A) <class 'str'><br>
(B) <class 'int'><br>
(C) <class 18><br>
(D) 18<br>

## Order of Operations


When more than one operator appears in an expression, the order of evaluation depends on the rules of precedence. Python follows the same precedence rules for its mathematical operators that mathematics does.

Parentheses have the highest precedence and can be used to force an expression to evaluate in the order you want. Since expressions in parentheses are evaluated first, 2 * (3-1) is 4, and (1+1)\**(5-2) is 8. 

You can also use parentheses to make an expression easier to read, as in (minute * 100) / 60, even though it doesn’t change the result.
Exponentiation has the next highest precedence, so 2\**1+1 is 3 and not 4, and 3*1\**3 is 3 and not 27. Can you explain why?

Multiplication and both division operators have the same precedence, which is higher than addition and subtraction, which also have the same precedence. So 2*3-1 yields 5 rather than 4, and 5-2*2 is 1, not 6.

Operators with the same precedence are evaluated from left-to-right. In algebra we say they are left-associative. So in the expression 6-3+2, the subtraction happens first, yielding 3. We then add 2 to get the result 5. If the operations had been evaluated from right to left, the result would have been 6-(3+2), which is 1.

> ***Note***
Due to some historical quirk, an exception to the left-to-right left-associative rule is the exponentiation operator \**. A useful hint is to always use parentheses to force exactly the order you want when exponentiation is involved:

In [37]:
print(2 ** 3 ** 2)     # the right-most ** operator gets done first!
print((2 ** 3) ** 2)   # use parentheses to force the order you want!


512
64


## Check your understanding

What is the value of the following expression:<br>
16 - 2 * 5 // 3 + 1<br>
(A) 14<br>
(B) 24<br>
(C) 3<br>
(D) 13.667<br>

 What is the value of the following expression:
2 \** 2 \** 3 \* 3<br>
(A) 768<br>
(B) 128<br>
(C) 12<br>
(D) 256<br>

## Reassignment

As we have mentioned previously, it is legal to make more than one assignment to the same variable. A new assignment makes an existing variable refer to a new value (and stop referring to the old value).



In [38]:
bruce = 5
print(bruce)
bruce = 7
print(bruce)


5
7


The first time bruce is printed, its value is 5, and the second time, its value is 7. The assignment statement changes the value (the object) that bruce refers to.

Here is what reassignment looks like in a reference diagram:

![](http://interactivepython.org/runestone/static/thinkcspy/_images/reassign1.png)

It is important to note that in mathematics, a statement of equality is always true. If a is equal to b now, then a will always equal to b. In Python, an assignment statement can make two variables refer to the same object and therefore have the same value. They appear to be equal. However, because of the possibility of reassignment, they don’t have to stay that way:

In [39]:
a = 5
b = a    # after executing this line, a and b are now equal
print(a, b)
a = 3    # after executing this line, a and b are no longer equal
print(a, b)


5 5
3 5


Line 4 changes the value of a but does not change the value of b, so they are no longer equal. We will have much more to say about equality in a later chapter.

> **Note** <br>
In some programming languages, a different symbol is used for assignment, such as <- or :=. The intent is that this will help to avoid confusion. Python chose to use the tokens = for assignment, and == for equality. This is a popular choice also found in languages like C, C++, Java, and C#.



## Check your understanding

After the following statements, what are the values of x and y?<br>
x = 15<br>
y = x<br>
x = 22<br>
(A) x is 15 and y is 15<br>
(B) x is 22 and y is 22<br>
(C) x is 15 and y is 22<br>
(D) x is 22 and y is 15<br>

## Updating Variables


One of the most common forms of reassignment is an update where the new value of the variable depends on the old. For example,

x = x + 1

This means get the current value of x, add one, and then update x with the new value. The new value of x is the old value of x plus 1. Although this assignment statement may look a bit strange, remember that executing assignment is a two-step process. First, evaluate the right-hand side expression. Second, let the variable name on the left-hand side refer to this new resulting object. The fact that x appears on both sides does not matter. The semantics of the assignment statement makes sure that there is no confusion as to the result.



In [40]:
x = 6        # initialize x
print(x)
x = x + 1    # update x
print(x)


6
7


If you try to update a variable that doesn’t exist, you get an error because Python evaluates the expression on the right side of the assignment operator before it assigns the resulting value to the name on the left. Before you can update a variable, you have to initialize it, usually with a simple assignment. In the above example, x was initialized to 6.

Updating a variable by adding 1 is called an increment; subtracting 1 is called a decrement. Sometimes programmers also talk about bumping a variable, which means the same as incrementing it by 1.



## Check your understanding

What is printed when the following statements execute?<br>
x = 12<br>
x = x - 1<br>
print(x)<br>
(A) 12<br>
(B) -1<br>
(C) 11<br>
(D) Nothing. An error occurs because x can never be equal to x - 1.<br><br>

What is printed when the following statements execute?<br>
x = 12<br>
x = x - 3<br>
x = x + 5<br>
x = x + 1<br>
print(x)<br>
(A) 12<br>
(B) 9<br>
(C) 15<br>
(D) Nothing. An error occurs because x cannot be used that many times in assignment statements.<br>

## Glossary
**assignment statement**<br>
A statement that assigns a value to a name (variable). To the left of the assignment operator, =, is a name. To the right of the assignment token is an expression which is evaluated by the Python interpreter and then assigned to the name. The difference between the left and right hand sides of the assignment statement is often confusing to new programmers. In the following assignment:

<pre>n = n + 1</pre>
n plays a very different role on each side of the =. On the right it is a value and makes up part of the expression which will be evaluated by the Python interpreter before assigning it to the name on the left.

**assignment token**<br>
= is Python’s assignment token, which should not be confused with the mathematical comparison operator using the same symbol.

**class**<br>
see data type below

**comment**<br>
Information in a program that is meant for other programmers (or anyone reading the source code) and has no effect on the execution of the program.

**data type**<br>
A set of values. The type of a value determines how it can be used in expressions. So far, the types you have seen are integers (int), floating-point numbers (float), and strings (str).

**decrement**<br>
Decrease by 1.

**evaluate**<br>
To simplify an expression by performing the operations in order to yield a single value.

**expression**<br>
A combination of operators and operands (variables and values) that represents a single result value. Expressions are evaluated to give that result.

**float**<br>
A Python data type which stores floating-point numbers. Floating-point numbers are stored internally in two parts: a base and an exponent. When printed in the standard format, they look like decimal numbers. Beware of rounding errors when you use floats, and remember that they are only approximate values.

**increment**<br>
Both as a noun and as a verb, increment means to increase by 1.
initialization (of a variable)
To initialize a variable is to give it an initial value. Since in Python variables don’t exist until they are assigned values, they are initialized when they are created. In other programming languages this is not the case, and variables can be created without being initialized, in which case they have either default or garbage values.

**int**<br>
A Python data type that holds positive and negative whole numbers.
integer division
An operation that divides one integer by another and yields an integer. Integer division yields only the whole number of times that the numerator is divisible by the denominator and discards any remainder.

**keyword**<br>
A reserved word that is used by the compiler to parse program; you cannot use keywords like if, def, and while as variable names.
modulus operator
Also called remainder operator or integer remainder operator. Gives the remainder after performing integer division.

**object**<br>
Also known as a data object (or data value). The fundamental things that programs are designed to manipulate (or that programmers ask to do things for them).

**operand**<br>
One of the values on which an operator operates.

**operator**<br>
A special symbol that represents a simple computation like addition, multiplication, or string concatenation.

**prompt string**<br>
Used during interactive input to provide the use with hints as to what type of value to enter.

**reference diagram**<br>
A picture showing a variable with an arrow pointing to the value (object) that the variable refers to. See also state snapshot.

**rules of precedence**<br>
The set of rules governing the order in which expressions involving multiple operators and operands are evaluated.

**state snapshot**<br>
A graphical representation of a set of variables and the values to which they refer, taken at a particular instant during the program’s execution.

**statement**<br>
An instruction that the Python interpreter can execute. So far we have only seen the assignment statement, but we will soon meet the import statement and the for statement.

**str**<br>
A Python data type that holds a string of characters.

**type conversion function**<br>
A function that can convert a data value from one type to another.

**value**<br>
A number or string (or other things to be named later) that can be stored in a variable or computed in an expression.

**variable**<br>
A name that refers to a value.

**variable name**<br>
A name given to a variable. Variable names in Python consist of a sequence of letters (a..z, A..Z, and _) and digits (0..9) that begins with a letter. In best programming practice, variable names should be chosen so that they describe their use in the program, making the program self documenting.

## Exercises

Evaluate the following numerical expressions in your head, then use the active code window to check your results:
<br>
5 \** 2<br>
9 * 5<br>
15 / 12<br>
12 / 15<br>
15 // 12<br>
12 // 15<br>
5 % 2<br>
9 % 5<br>
15 % 12<br>
12 % 15<br>
6 % 6<br>
0 % 7<br>


In [41]:
print("Test here!")

Test here!


What is the order of the arithmetic operations in the following expression. Evaluate the expression by hand and then check your work.
2 + (3 - 1) * 10 / 5 * (2 + 3)

Many people keep time using a 24 hour clock (11 is 11am and 23 is 11pm, 0 is midnight). If it is currently 13 and you set your alarm to go off in 50 hours, it will be 15 (3pm). Write a Python program to solve the general version of the above problem. Ask the user for the time now (in hours), and then ask for the number of hours to wait for the alarm. Your program should output what the time will be on the clock when the alarm goes off.



It is possible to name the days 0 through 6 where day 0 is Sunday and day 6 is Saturday. If you go on a wonderful holiday leaving on day number 3 (a Wednesday) and you return home after 10 nights. Write a general version of the program which asks for the starting day number, and the length of your stay, and it will tell you the number of day of the week you will return on.




Take the sentence: All work and no play makes Jack a dull boy. Store each word in a separate variable, then print out the sentence on one line using print.



Add parenthesis to the expression 6 * 1 - 2 to change its value from 4 to -6.



The formula for computing the final amount if one is earning compound interest is given on Wikipedia as

formula for compound interest
Write a Python program that assigns the principal amount of 10000 to variable P, assign to n the value 12, and assign to r the interest rate of 8% (0.08). Then have the program prompt the user for the number of years, t, that the money will be compounded for. Calculate and print the final amount after t years.

RunShow FeedbackShow CodeShow CodeLensCode Coach
(ex_2_7)
Write a program that will compute the area of a circle. Prompt the user to enter the radius and print a nice message back to the user with the answer.
RunShow FeedbackShow CodeShow CodeLensCode Coach
(ex_2_8)
Question
Answer
Discussion
Write a program that will compute the area of a rectangle. Prompt the user to enter the width and height of the rectangle. Print a nice message with the answer.

RunShow FeedbackShow CodeShow CodeLensCode Coach
(ex_2_9)
Write a program that will compute MPG for a car. Prompt the user to enter the number of miles driven and the number of gallons used. Print a nice message with the answer.
RunShow FeedbackShow CodeShow CodeLensCode Coach
(ex_2_10)
Question
Answer
Discussion
Write a program that will convert degrees celsius to degrees fahrenheit.

RunShow FeedbackShow CodeShow CodeLensCode Coach
(ex_2_11)
Write a program that will convert degrees fahrenheit to degrees celsius.
RunShow FeedbackShow CodeShow CodeLensCode Coach
